std.PriorityQueue: Convert to an 'unmanaged'-style API

saltzm opened the pull request:
Resolves: #21432.

Removed the allocator field from the PriorityQueue struct and adapted functions to take an allocator where needed.

Updated tests accordingly.

If these changes look good, I'll make corresponding changes to PriorityDequeue and include them in a follow-up patch.

saltzm commented on lib/std/priority_queue.zig:
```diff
@@ -18,33 +18,26 @@ pub fn PriorityQueue(comptime T: type, comptime Context: type, comptime compareF
+        items: []T = undefined,
+        /// The capacity of the queue. This may be read directly, but must not
+        /// be modified directly.
+        capacity: usize = 0,
```
I made this change to match multi_array_list.zig and because this is the only standard library container I've seen with a `capacity()` function. It's not that I think the idea of a `capacity()` function is bad, just that it's possibly nonstandard.
 
In general it seems like there's not a fixed standard for getting the current size of a container or for getting its capacity, and oftentimes you have to access members directly - would be nice to standardize this across containers down the line.

wooster0 replied:
No, it was definitely bad. What you did is much better

xdBronch commented on lib/std/priority_queue.zig:
```diff
@@ -18,33 +18,26 @@ pub fn PriorityQueue(comptime T: type, comptime Context: type, comptime compareF
         context: Context,
 
-        /// Initialize and return a priority queue.
-        pub fn init(allocator: Allocator, context: Context) Self {
```
i think that this should still use an init function rather than default values, if you were to for example initialize it with `.{ .capacity = 5, .context = .{} }` this would be invalid as the `items` field wouldnt correctly fit the given capacity. see <https://ziglang.org/documentation/master/#Faulty-Default-Field-Values>

xdBronch replied:
this rule/guideline/whatever you might call it regarding default values is fairly new, im not sure why multiarraylist is still using default values but i think it really shouldnt, a few other types have started moving in the same direction https://github.com/ziglang/zig/commit/0b9fccf508dc85fa522947d1cf6ff84f78f2dcb4

andrewrk commented on lib/std/priority_queue.zig:
```diff
@@ -18,33 +18,26 @@ pub fn PriorityQueue(comptime T: type, comptime Context: type, comptime compareF
-        items: []T,
-        cap: usize,
-        allocator: Allocator,
+        items: []T = undefined,
```
default-initializing a field to `undefined` is a code smell. Almost always would be better with an init function or init value.

saltzm replied:
Makes sense, I agree and as I mentioned in my other comment would normally do an `init()` function - I was just copying from [MultiArrayList](https://github.com/ziglang/zig/blob/41330c96aebddbf26d8fdc0725e7483476175601/lib/std/multi_array_list.zig#L22) in case that was somehow preferred. 

I'll change it to use an `init` function.

saltzm commented on lib/std/priority_dequeue.zig:
```diff
@@ -21,39 +21,41 @@ pub fn PriorityDequeue(comptime T: type, comptime Context: type, comptime compar
-        pub fn deinit(self: Self) void {
-            self.allocator.free(self.items);
+        pub fn deinit(self: Self, allocator: std.mem.Allocator) void {
+            if (self.items.len > 0) {
```
This isn't strictly necessary since `free` checks for 0-length slices but it still seemed odd to me to intentionally free something that was not a result of a call to `alloc` so I added this check.

saltzm commented:
I added back the `PriorityQueue.init()` function and changed PriorityDequeue to be unmanaged. 

I also noticed that what PriorityQueue was doing with the `capacity` field was a bit odd and did not match PriorityDequeue. It was tracking its current size by modifying the `items.len` field, and then reconstituting the "allocated slice" by taking `self.items.ptr[0..self.capacity]`. This seemed a little odd to me so I changed it to match the way PriorityDequeue is implemented, tracking the length separately and having `items` always be the full backing array/slice of data. The capacity is then accessed using `queue.items.len`.

PriorityQueue and PriorityDequeue still have a `count()` function that returns the size of the queue, and PriorityDequeue still has a `capacity()` function. Before changing it I wanted to ask what y'all thought the best API would be. In my mind there are two options:
# Provide both `count()` and `capacity()` functions
# Provide neither functions. The user gets the count by doing `queue.len` and gets the capacity by doing `queue.items.len`

The latter I think matches the rest of the standard library more, though as a user I have sometimes found it a little odd that I need to access fields directly for this information rather than calling some consistent function.

What do y'all think the best move would be here?

andrewrk commented:
> I also noticed that what PriorityQueue was doing with the `capacity` field was a bit odd and did not match PriorityDequeue. It was tracking its current size by modifying the `items.len` field, and then reconstituting the "allocated slice" by taking `self.items.ptr[0..self.capacity]`. This seemed a little odd to me so I changed it to match the way PriorityDequeue is implemented, tracking the length separately and having `items` always be the full backing array/slice of data. The capacity is then accessed using `queue.items.len`.

This is probably motivated by what ArrayList does and I think people generally agree (myself included) that it's preferred over the alternative because it means that your `items` field is always safe to access directly, and you have to be a bit more intentional to access uninitialized memory.

saltzm commented:
> This is probably motivated by what ArrayList does and I think people generally agree (myself included) that it's preferred over the alternative because it means that your items field is always safe to access directly, and you have to be a bit more intentional to access uninitialized memory.

Okay, I put it in its own commit so it's easy to undo. I'll undo it.

andrewrk commented on lib/std/priority_queue.zig:
```diff
@@ -19,32 +19,34 @@ pub fn PriorityQueue(comptime T: type, comptime Context: type, comptime compareF
         items: []T,
-        cap: usize,
-        allocator: Allocator,
+        /// The capacity of the queue. This may be read directly, but must not
+        /// be modified directly.
```
* "the capacity of the queue" is 100% redundant with the name of the data structure and the field. Delete it.
* documentation should be descriptive, not prescriptive. Instead of telling the person what they should/shouldn't do, may or may not do, just document how it works.

```suggestion
        /// Tracks the allocated slice of memory when combined with `items.ptr`.
```

You can see how this comment is more useful, and the programmer can infer that they can read it directly, and that modifying it directly is in fact possible if they also modify `items` appropriately.

Missing documentation is acceptable. Documentation that violates the redundancy principle or descriptive principle is not. When in doubt, delete the documentation.

saltzm replied:
I'm happy to change the comment and accept your suggestion. I do see how yours is more descriptive. 

I do however want to at least express my disagreement with this part:
> the programmer can infer that they can read it directly, and that modifying it directly is in fact possible if they also modify items appropriately.

I don't agree that the programmer either can, or should have to, infer which fields of a struct they are able to access directly. 

First, as someone who has had to navigate the standard library, I've found it a bit tedious that there's no standard way to get the capacity or current size of a container. I think for `ArrayList` I had to go read the tests to see that I'm supposed to access `items.len` directly to get the size of the container. While for another container (say, `PriorityDequeue`) you would have to do something different (use the `count()` function or `len` field). In the absence of standardization, documentation on a field to say that it's appropriate to access the field directly would be helpful so that I don't need to understand the implementation of the data structure just to get the size or capacity of it. I know that the standard library is still a work in progress and I think last I saw there's a plan to have a bit of a cleaning up/standardization later (is that true?) but I would hope that in the final incarnation, containers would have a consistent and documented way to fetch their size and capacity. And if that's via direct field access, I would hope that that would be documented. (My preference would be via a function with a name that isn't a noun - e.g. `getCapacity()` rather than `capacity()` which can conflict with variable names.)

Second, once back-compatibility starts to matter, it could be a big issue to have users accessing fields directly. Say a user is accessing the `items` field of `ArrayList` directly in order to obtain the current size of the list by doing `items.len`. If later on `items` were changed to store the allocated slice rather than the slice of current items, that would silently change the functionality of that users code, which would be a big problem if there were any promise of back-compatibility. (I know that isn't a change that's likely to be made, just an example :)) 

In other words, as soon as you say that users should directly access fields of a struct, you're effectively saying that fields become part of the API and now need to be considered as part of upgrades/deprecation/back-compatibility. This then limits your ability to swap out the internals of the data structure without impacting user code. If that is a choice that you're willing to make, then it seems like it would be a hard requirement to document which fields are expected to be accessed directly so that users can rely on that for stability and so maintainers know which fields need to undergo a deprecation process if they are removed or changed somehow. Otherwise, without this kind of documentation, I think all fields would need to be considered as public and unable to be removed/have their behavior changed without a deprecation process, since you're essentially saying they are all fair game to be accessed as the user sees fit.

andrewrk replied:
There are no private fields in Zig. All fields are public and part of the API, and all APIs must take this into account.

saltzm replied:
Got it. I knew there was no support in the language for private fields but did not realize that the intent was also for all the fields to be conceptually part of the API. Thanks for explaining.

dannyoler commented:
Why is the queue's pop api called `remove()` instead of `pop()` like other zig list structs?

alexrp commented:
Would you mind resolving the merge conflicts here?

andrewrk commented:
closing abandoned PR
